src/spans.rs

```diff
@@ -0,0 +1,355 @@
+//! Source code spans.
+use std::{
+    collections::{BTreeSet, VecDeque},
+    marker::PhantomData,
+    mem,
+};
+
+use crate::{Attribute, Doctype, Emitter, EndTag, Error, Reader, StartTag, Token};
+
+type Span = std::ops::Range<usize>;
+
+/// A trait to be implemented by readers that track their own position.
+pub trait GetPos {
+    /// Returns the byte index of the current position.
+    fn get_pos(&self) -> usize;
+}
+
+/// Wraps a [`Reader`] so that it implements [`GetPos`].
+pub struct PosTracker<R> {
+    /// The wrapped reader.
+    pub reader: R,
+    /// The current position.
+    pub position: usize,
+}
+
+impl<R> GetPos for PosTracker<R> {
+    fn get_pos(&self) -> usize {
+        self.position
+    }
+}
+
+impl<R: Reader> Reader for PosTracker<R> {
+    type Error = R::Error;
+
+    fn read_char(&mut self) -> Result<Option<char>, Self::Error> {
+        match self.reader.read_char()? {
+            Some(char) => {
+                self.position += char.len_utf8();
+                Ok(Some(char))
+            }
+            None => Ok(None),
+        }
+    }
+
+    fn try_read_string(&mut self, s: &str, case_sensitive: bool) -> Result<bool, Self::Error> {
+        match self.reader.try_read_string(s, case_sensitive)? {
+            true => {
+                self.position += s.len();
+                Ok(true)
+            }
+            false => Ok(false),
+        }
+    }
+}
+
+/// The default implementation of [`crate::Emitter`], used to produce ("emit") tokens.
+pub struct SpanEmitter<R> {
```

Comment:
Can't tnis again wrap the default emitter, or maybe even arbitrary ones? this seems like a ton of duplication

I could imagine that for example:

```
impl<E: Emitter, R: Reader> Emitter<R> for SpanEmitter<E, R> {
    type Token = (Span, E::Token);
}
```

(as an interface, I am not sure at all whether this can be implemented)

the reason i am grilling you on this is because it would be nice to get Spans for custom emitters. Usecase for custom emitters are ones that use a custom string type to store name tags, or that filter out uninteresting data.

im also curious whether there needs to be a separate emitter type at all, or whether the defaultemitter can implement GetPos while still being generic over whether it emits any Spans. After all the defaultemitter already maintains a usize for the position

Reply:
> I could imagine `type Token = (Span, E::Token);`

This would only provide spans for the whole start tag but not for tag names and attribute names/values, which I require for my linter.

> this seems like a ton of duplication

Valid point. My `SpanEmitter` was just a copy of your `DefaultEmitter` with some additions. In my latest commit I removed the `SpanEmitter` and instead made the necessary changes in the `DefaultEmitter`. For this I introduce a `Span` trait that takes care of the actual tracking and is implemented for `()` (for which it doesn't do anything) and `Range<usize>` for which it retrieves the current position from the reader / advances the span when a string is pushed.

>the reason i am grilling you on this is because it would be nice to get Spans for custom emitters

That's now possible.